### PR TITLE
Fix test delegation approval

### PR DESCRIPTION
**Pull request by 2byrds**
Delegation Approval enpoint added.
The formatting for test_aiding was updated but the only significant change is the test_identifier_delegation_end test.

### CODE REVIEW

**Comment by 2byrds** [comment]
Depends on https://github.com/WebOfTrust/keripy/pull/789/files

**Comment by 2byrds** [comment]
Per the discussion on #247 I'm going to adjust the endpoint for the approval on this PR.

**Comment by 2byrds** [comment]
Updated per @lenkan suggestion to use a separate endpoint and a PUT instead of POST

**Comment by lenkan** [comment]
Cross posting my comment on #247 for visibility 

> What is the process?
> 
> 1. Delegatee create delegated inception event `dip`
> 2. Delegator receives the inception event details out of band
> 3. Delegator create interaction event with delegation information in anchor data.
> 4. Delegator approves delegation by posting the same interaction event again?
> 
> What is the risk with having step 4 processed automatically?
> 
> If it is not possible to do automatically, can we add a flag to step 3 so it becomes
> 
> ```
> POST /identifiers/name/events
> { 
>    ixn: [ anchor data ], 
>    approveDelegation: true 
> }
> ```
> 
> Another option if we need to explicitly call the server again, can we move it to another endpoint? What is it that is being created or modified here? Is it a "pending delegation request"?
> 
> Note: I am not saying any of my suggestion are objectively better, just trying to trigger discussion about how we design the HTTP API of KERIA so that we can improve it as we move forward.

I think from the dev meeting today it is clear that the desired approach is to _not_ have to do step 4 as a separate API call, but rather as a part of the call that posts the interaction event.

**Comment by 2byrds** [comment]
> Cross posting my comment on #247 for visibility 
> 
> > What is the process?
> > 
> > 1. Delegatee create delegated inception event `dip`
> > 2. Delegator receives the inception event details out of band
> > 3. Delegator create interaction event with delegation information in anchor data.
> > 4. Delegator approves delegation by posting the same interaction event again?
> > 
> > What is the risk with having step 4 processed automatically?
> > 
> > If it is not possible to do automatically, can we add a flag to step 3 so it becomes
> > 
> > ```
> > POST /identifiers/name/events
> > { 
> >    ixn: [ anchor data ], 
> >    approveDelegation: true 
> > }
> > ```
> > 
> > Another option if we _absolutely need_ to do do another call to the server, we can move it to another endpoint? What is it that is being created or modified here? Is it a "pending delegation request"?
> > 
> > Note: I am not saying any of my suggestion are objectively better, just trying to trigger discussion about how we design the HTTP API of KERIA so that we can improve it as we move forward.
> 
> I think from the dev meeting today it is clear that the desired approach is to _not_ have to do step 4 as a separate API call, but rather as a part of the call that posts the interaction event.

I agree, let's hear from @pfeairheller and I'll adjust the PR accordingly

File: src/keria/app/aiding.py [hunk 1]
```diff
@@ -245,6 +248,110 @@ def anchorSeals(agent, ixn):
...
+        gatePre = ixn['a'][0]['i']
+        gateSaid = ixn['a'][0]['d']
+
+        for (pre, sn), dig in hab.db.delegables.getItemIter():
+            if pre == gatePre:
+                seqner = coring.Seqner(sn=serder.sn)
+                couple = seqner.qb64b + serder.saidb
+                dgkey = dbing.dgKey(coring.Saider(qb64=gatePre).qb64b, coring.Saider(qb64=gateSaid).qb64b)
+                hab.db.setAes(dgkey, couple)  # authorizer event seal (delegator/issuer)
+
```
Comment (pfeairheller) [inline src/keria/app/aiding.py:349]: You need to parse the DIP event to get it into your kevers.  Just creating the Aes isn't sufficient.

**Review by pfeairheller — CHANGES_REQUESTED** [review]
A few concerns here...

First, the history on this PR seems odd as it is bringing in some of my commits from 9 days ago.

Second, I don't think REST likes a PUT on a plural endpoint, PUTs require a resource afaik.

Third, I'm not sure why there is the waterfall of excepts at the bottom of the on_put, none of those exceptions can happen in that try block.

Fourth, we have a "delegating.py" file where this code should belong and the endpoint should be "delegation" specific, not "approvals"

Finally (and most importantly) I don't think this is the right approach.  If you look at credentialing.py, there are endpoints that do credential specific things that ALSO require an anchoring event.  That is the pattern that should be followed for this.   A delegation endpoint that accepts either an IXN or ROT that is processed by the IdentifierResource, then a long running operation is returned.  A new cue can be submitted to a Doer that then checks the delegables escrow, creates the anchor seal attachment and parses the event and when it is actually committed then declares the operation "done".  So you'll need a new long running operation type with code in the Monitor to check for this to be "done" by looking for the successful completion of the delegated event.

My suggestion is to close this PR and open a new one that takes another approach.

File: setup.py [hunk 2]
```diff
@@ -76,7 +76,7 @@
     python_requires='>=3.12.2',
     install_requires=[
         'hio>=0.6.12',
-        'keri==1.2.0-dev4',
+        'keri @ git+https://github.com/WebOfTrust/keripy.git@main',
```
[the remark this answers is not in the document]
Comment (2byrds) [reply setup.py:79]: agree

File: src/keria/app/aiding.py [hunk 3]
```diff
@@ -28,6 +28,9 @@ def loadEnds(app, agency, authn):
     groupEnd = AgentResourceEnd(agency=agency, authn=authn)
     app.add_route("/agent/{caid}", groupEnd)
 
+    approvalEnd = IdentifierApprovalEnd()
+    app.add_route("/identifiers/{name}/approvals", approvalEnd)
+
```
[the remark this answers is not in the document]
Comment (2byrds) [reply src/keria/app/aiding.py:33]: sounds good thank you

File: src/keria/app/aiding.py [hunk 4]
```diff
@@ -245,6 +248,110 @@ def anchorSeals(agent, ixn):
...
+        for (pre, sn), dig in hab.db.delegables.getItemIter():
+            if pre == gatePre:
+                seqner = coring.Seqner(sn=serder.sn)
+                couple = seqner.qb64b + serder.saidb
+                dgkey = dbing.dgKey(coring.Saider(qb64=gatePre).qb64b, coring.Saider(qb64=gateSaid).qb64b)
+                hab.db.setAes(dgkey, couple)  # authorizer event seal (delegator/issuer)
+
+                op = agent.monitor.submit(hab.kever.prefixer.qb64, longrunning.OpTypes.done,
+                                        metadata=dict(response=serder.ked))
+                return op
```
[the remark this answers is not in the document]
Comment (2byrds) [reply src/keria/app/aiding.py:352]: sounds good thank you

File: src/keria/app/aiding.py [hunk 5]
```diff
@@ -245,6 +248,110 @@ def anchorSeals(agent, ixn):
...
+            op = self.approveDelegation(agent, name, body)
+        else:
+            raise falcon.HTTPBadRequest(title="invalid request",
+                                        description=f"required field 'approveDelegation' missing from request")
+
+        try:
+            rep.status = falcon.HTTP_200
+            rep.content_type = "application/json"
+            rep.data = op.to_json().encode("utf-8")
+        except (kering.AuthError, ValueError) as e:
```
[the remark this answers is not in the document]
Comment (2byrds) [reply src/keria/app/aiding.py:295]: i'll double check this. i wouldn't have added them without experiencing something during testing. but i've forgotten the exact context and will revalidate and/or remove.

File: src/keria/app/aiding.py (hunk 1 above)
Comment (2byrds) [reply src/keria/app/aiding.py:349]: The dip event is waiting in escrows to be processed but the one dependency needed is Aes. Based on our previous discussions I assumed this was the proper solution to allow the processing. I'll adjust it however you prefer.

**Comment by 2byrds** [comment]
Closing now that review comments addressed in new PR https://github.com/WebOfTrust/keria/pull/250

**Pull request closed without merging**